### PR TITLE
add slack notification system for failed event deliveries

### PR DESCRIPTION
**Pull request by danvixent**

### CODE REVIEW

**Review by jirevwe — APPROVED** [review]

**Review by ogbanugot — APPROVED** [review]

File: worker/task/process_event_delivery.go [hunk 1]
```diff
@@ -299,17 +291,65 @@ func ProcessEventDelivery(appRepo datastore.ApplicationRepository, eventDelivery
 	}
 }
 
-func sendEmailNotification(m *datastore.EventDelivery, g *datastore.Group, s *smtp.SmtpClient, status datastore.EndpointStatus) error {
-	email := m.AppMetadata.SupportEmail
+func sendEndpointReactivationNotification(ctx context.Context, appRepo datastore.ApplicationRepository, eventDelivery *datastore.EventDelivery, g *datastore.Group, smtpCfg *config.SMTPConfiguration, status datastore.EndpointStatus) {
+	app, err := appRepo.FindApplicationByID(ctx, eventDelivery.AppMetadata.UID)
+	if err != nil {
+		log.WithError(err).Error("failed to fetch application")
```
Comment (Subomi) [inline worker/task/process_event_delivery.go:297]: @danvixent We shouldn't be logging here. Just return the error.

File: worker/task/process_event_delivery.go [hunk 2]
```diff
@@ -299,17 +291,65 @@ func ProcessEventDelivery(appRepo datastore.ApplicationRepository, eventDelivery
...
+		TargetURL:      eventDelivery.EndpointMetadata.TargetURL,
+		EndpointStatus: string(status),
+	}
+
+	for i := range app.NotificationChannels {
+		SendNotification(ctx, &app.NotificationChannels[i], n, smtpCfg)
+	}
+}
+
+func SendNotification(ctx context.Context, nc *datastore.NotificationChannel, n *notification.Notification, smtpCfg *config.SMTPConfiguration) {
```
Comment (Subomi) [inline worker/task/process_event_delivery.go:332]: @danvixent Is there a reason this is an exported function?

File: smtp/smtp.go [hunk 3]
```diff
@@ -105,11 +104,14 @@ func (s *SmtpClient) setHeaders(email string) *gomail.Message {
...
+	if !IsStringEmpty(s.replyTo) {
 		m.SetHeader("Reply-To", s.replyTo)
 	}
 
 	m.SetHeader("Subject", NotificationSubject)
 
 	return m
 }
+
+// IsStringEmpty checks if the given string s is empty or not
```
Comment (Subomi) [inline smtp/smtp.go:116]: @danvixent Why? This function already exists in `util`

File: datastore/models.go [hunk 4]
```diff
@@ -72,12 +78,13 @@ const (
...
-	GroupID      string             `json:"group_id" bson:"group_id"`
-	Title        string             `json:"name" bson:"title"`
-	SupportEmail string             `json:"support_email" bson:"support_email"`
-	IsDisabled   bool               `json:"is_disabled" bson:"is_disabled"`
+	ID                   primitive.ObjectID    `json:"-" bson:"_id"`
+	UID                  string                `json:"uid" bson:"uid"`
+	GroupID              string                `json:"group_id" bson:"group_id"`
+	Title                string                `json:"name" bson:"title"`
+	SupportEmail         string                `json:"support_email" bson:"support_email"`
+	NotificationChannels []NotificationChannel `json:"notification_channels,omitempty" bson:"notification_channels"`
```
[the remark this answers is not in the document]
Comment (Subomi) [reply datastore/models.go:80]: @danvixent Please implement this spec. The `NotificationChannel` unnecessarily complicates the server package when we can marshal and rely on `govalidate` for validation.

**Review by Subomi — CHANGES_REQUESTED** [review]
@danvixent FYI

File: datastore/models.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (danvixent) [reply datastore/models.go]: @Subomi done, alongside all the other suggestions

File: worker/task/notification.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	eventDelivery *datastore.EventDelivery,
+	g *datastore.Group,
+	smtpCfg *config.SMTPConfiguration,
+	status datastore.EndpointStatus,
+	failure bool,
+) {
+	app, err := appRepo.FindApplicationByID(ctx, eventDelivery.AppMetadata.UID)
+	if err != nil {
+		log.WithError(err).Error("failed to fetch application")
+		return
```
Comment (Subomi) [inline worker/task/notification.go:29]: @danvixent This should be `return err` and we shouldn't be logging here at all only return err.

Comment (danvixent) [reply worker/task/notification.go:29]: The ProcessEventDelivery function's flow cannot be disrupted by the error here, that is why I think it's best to log here.

What do you think?

**Pull request merged**